src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs

```diff
@@ -0,0 +1,465 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    ///  Represents a JSON object.
+    /// </summary>
+    public sealed class JsonObject : JsonNode, IEnumerable<KeyValuePair<string, JsonNode>>
+    {
+        private readonly Dictionary<string, JsonNode> _dictionary;
+        private readonly DuplicatePropertyNameHandling _duplicatePropertyNameHandling;
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonObject"/> class representing the empty object.
+        /// </summary>
+        /// <param name="duplicatePropertyNameHandling">Specifies the way of handling duplicate property names.</param>
+        /// <exception cref="ArgumentException">
+        ///   Provided manner of handling duplicates does not exist.
+        /// </exception>
+        public JsonObject(DuplicatePropertyNameHandling duplicatePropertyNameHandling = default)
+        {
+            if(!Enum.IsDefined(typeof(DuplicatePropertyNameHandling), duplicatePropertyNameHandling))
```

Comment:
Consider something like this for a quicker check (as long as the tests pass):
[sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGue/u9BjAB2AV3wMAImIAOAGwCWYbBhgAFKBFkwoGAJ4A5bPhgAJbCIAmSkQHMBHRzwYAlGAuxgYDALwMABjRnHgBJOxFoH39yYOoXFwBRKC0oPwZSOJcAX0dHIWAICHkGAFlyAAoZBWVVDS0dPSMTc0sbRXsGKzklFTVNbV0DY1MLa1s7dKsYADNsMXkMAEoneITBAHYGAEJE8XwAOlDcKVmOmCsKgx0IGaqe2v6GoebRtomltC6HvvrBppGrXGHTsSwA3I5cmtuM4CkUSqVSPcar8Bo1hi0xu1Ot0UXU0S9AViJlNZvNFitOND1sQthUKmIOstcb18c8AZj3iCGAA+BgMplLaqsp7/DFvYH2A7JVLgyE0bJAA===)

```C#
((uint)duplicatePropertyNameHandling > (uint)DuplicatePropertyNameHandling.Error)
```

Reply:
I chose `Enum.IsDefined`, because enums can have not subsequent values and start from something different than 0. But if we wouldn't want to change it in future, above check should work for current version.

Reply:
Good point. However, we have similar checks elsewhere within the JSON API implementation, and since the enum is public, adding to it is an API change which would need to go through review anyway. Modifying the values would be considered breaking so we likely wouldn't do that.
This is why I had mentioned adding the test cases, which should guard against evolution of the enum. If someone was to add enum values, the test would fail, which would highlight the need to update this condition.